addons/website_slides/models/slide_slide.py

```diff
@@ -335,58 +361,116 @@ def _compute_user_membership_id(self):
+        for slide in self:
+            base_url = request and request.httprequest.url_root
             if not base_url:
-                base_url = record.get_base_url()
+                base_url = slide.get_base_url()
             if base_url[-1] == '/':
                 base_url = base_url[:-1]
-            if record.datas and (not record.document_id or record.slide_type in ['document', 'presentation']):
-                slide_url = base_url + url_for('/slides/embed/%s?page=1' % record.id)
-                record.embed_code = '<iframe src="%s" class="o_wslides_iframe_viewer" allowFullScreen="true" height="%s" width="%s" frameborder="0"></iframe>' % (slide_url, 315, 420)
-            elif record.slide_type == 'video' and record.document_id:
-                if not record.mime_type:
-                    # embed youtube video
-                    query = urls.url_parse(record.url).query
-                    query = query + '&theme=light' if query else 'theme=light'
-                    record.embed_code = '<iframe src="//www.youtube-nocookie.com/embed/%s?%s" allowFullScreen="true" frameborder="0"></iframe>' % (record.document_id, query)
+
+            embed_code = False
+            if slide.slide_type == 'video':
+                if slide.video_source_type == 'youtube':
+                    query_params = urls.url_parse(slide.video_url).query
+                    query_params = query_params + '&theme=light' if query_params else 'theme=light'
+                    embed_code = '<iframe src="//www.youtube-nocookie.com/embed/%s?%s" allowFullScreen="true" frameborder="0"></iframe>' % (slide.video_youtube_id, query_params)
+                elif slide.video_source_type == 'google_drive':
+                    embed_code = '<iframe src="//drive.google.com/file/d/%s/preview" allowFullScreen="true" frameborder="0"></iframe>' % (slide.video_google_drive_id)
+            elif slide.slide_type in ['infographic', 'document'] and slide.source_type == 'external' and slide.document_google_drive_id:
+                embed_code = '<iframe src="//drive.google.com/file/d/%s/preview" allowFullScreen="true" frameborder="0"></iframe>' % (slide.document_google_drive_id)
+            elif slide.slide_type == 'document' and slide.source_type == 'local_file':
+                slide_url = base_url + url_for('/slides/embed/%s?page=1' % slide.id)
+                embed_code = '<iframe src="%s" class="o_wslides_iframe_viewer" allowFullScreen="true" height="%s" width="%s" frameborder="0"></iframe>' % (slide_url, 315, 420)
+
+            slide.embed_code = embed_code
+
+    @api.depends('video_url')
+    def _compute_video_source_type(self):
+        for slide in self:
+            video_source_type = False
+            youtube_match = re.match(self.YOUTUBE_VIDEO_ID_REGEX, slide.video_url) if slide.video_url else False
+            if youtube_match:
+                if youtube_match and len(youtube_match.groups()) == 2 and len(youtube_match.group(2)) == 11:
```

Comment:
Mmmh ? would maybe extract to a sub tool method, to ease method understanding, like
```
if self._slide_video_is_youtube(self.video_url):
    video_source_type = 'youtube'
 elif self._slide_video_is_google(self.video_url):
    video_source_type = 'google_drive'
```